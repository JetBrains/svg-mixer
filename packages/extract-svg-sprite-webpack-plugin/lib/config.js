const fs = require('fs');

const mixer = require('svg-mixer');

const RuntimeGenerator = require('./utils/runtime-generator');

/**
 * @typedef {Object} ExtractSvgSpritePluginConfig
 * @property {string|function(path, query)} symbolId='[name]'
 * @property {string|function(path, query)} filename='sprite.svg'
 * @property {boolean} emit=true
 * @property {string} publicPath
 * @property {RuntimeGenerator} runtimeGenerator
 * @property {string} selector=null
 * @property {string} spriteType 'classic' | 'stack'
 * @property {mixer.Sprite} spriteClass
 * @property {mixer.SpriteSymbol} symbolClass
 */

module.exports = {
  CSS_LOADER_PATH: require.resolve('./css-loader'),
  LOADER_PATH: require.resolve('./loader'),
  NAMESPACE: fs.realpathSync(__dirname),

  /**
   * @return {ExtractSvgSpritePluginConfig}
   */
  plugin: {
    emit: true,
    filename: 'sprite.svg',
    publicPath: undefined,
    runtimeGenerator: RuntimeGenerator,
    selector: '',
    spriteClass: mixer.Sprite,
    spriteConfig: {
      usageIdSuffix: '-usage'
    },
    spriteType: mixer.Sprite.TYPE,
    symbolClass: mixer.SpriteSymbol,
    symbolId: '[name]'
  }
};
