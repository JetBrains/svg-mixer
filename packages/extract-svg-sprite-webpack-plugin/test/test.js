const webpack = require('webpack');
const memoryFs = require('memory-fs');
const { createCompiler, getAssets } = require('svg-mixer-test').webpack;

it('should works with extracted CSS', async () => {
  const compiler = createCompiler(
    require('./extract-css/webpack.config'),
    { webpack, memoryFs }
  );
  const compilation = await compiler.run();

  getAssets(compilation).forEach(({ name, content }) => {
    expect(content).toMatchSnapshot(name);
  });
});

it('should share sprite data with html-webpack-plugin', async () => {
  const compiler = createCompiler(
    require('./html-webpack-plugin-inline-sprite/webpack.config'),
    { webpack, memoryFs }
  );
  const compilation = await compiler.run();

  getAssets(compilation).forEach(({ name, content }) => {
    expect(content).toMatchSnapshot(name);
  });
});

it('should warn when symbols with duplicate ids exists', async () => {
  const compiler = createCompiler(
    require('./symbols-with-duplicate-ids/webpack.config'),
    { webpack, memoryFs }
  );

  expect((await compiler.run()).warnings).toHaveLength(1);
});
