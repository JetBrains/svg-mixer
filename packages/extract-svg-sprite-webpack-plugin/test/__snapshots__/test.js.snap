// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extract-svg-sprite-webpack-plugin should share sprite data with html-webpack-plugin: index.html 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"UTF-8\\">
  <title>Webpack App</title>
</head>
<body>

  <div style=\\"position: absolute; width: 0; height: 0; overflow: hidden;\\">
    <svg width=\\"273.4\\" height=\\"312.2\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"><defs><symbol viewBox=\\"0 0 70 70\\" id=\\"appcode\\"><path fill=\\"#247CE6\\" d=\\"M59.2 55L70 26.3 37.3 19 33 31.2z\\" /><path fill=\\"url(#appcode_a)\\" d=\\"M70 56.1L53.5 70 20.6 61.5 29.4 45.5z\\" /><path fill=\\"url(#appcode_b)\\" d=\\"M9.7 34.8L1.3 10.8 38.4 0 35.4 31z\\" /><path fill=\\"url(#appcode_c)\\" d=\\"M61.1 40.5L50.7 22.7 50.9 22.6 38.4 0 0 41.5 0 70 69.8 56.1z\\" /><g><path fill=\\"#000\\" d=\\"M13.4 13.4H56.6V56.6H13.4z\\" /><path fill=\\"#FFF\\" d=\\"M17.4 48.5H33.599999999999994V51.2H17.4z\\" /><path d=\\"M24.5,19h3.6l7.6,17.9h-4.1l-1.6-4h-7.5l-1.6,4h-4L24.5,19z M28.6,29.4l-2.4-5.8l-2.4,5.8H28.6z\\" fill=\\"#FFF\\" /><path d=\\"M34.7,28.1L34.7,28.1c0-5.2,3.9-9.4,9.4-9.4c3.4,0,5.4,1.1,7.1,2.8l-2.5,2.9c-1.4-1.3-2.8-2-4.6-2 c-3,0-5.2,2.5-5.2,5.6V28c0,3.1,2.1,5.7,5.2,5.7c2.1,0,3.3-0.8,4.7-2.1l2.5,2.5c-1.9,2-3.9,3.2-7.4,3.2 C38.6,37.3,34.7,33.2,34.7,28.1\\" fill=\\"#FFF\\" /></g></symbol><linearGradient id=\\"appcode_a\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"20.617\\" y1=\\"57.755\\" x2=\\"70\\" y2=\\"57.755\\"><stop offset=\\".194\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><linearGradient id=\\"appcode_b\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"1.274\\" y1=\\"17.415\\" x2=\\"38.41\\" y2=\\"17.415\\"><stop offset=\\".194\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><linearGradient id=\\"appcode_c\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"45.876\\" y1=\\"72.222\\" x2=\\"11.197\\" y2=\\"23.824\\"><stop offset=\\".091\\" stop-color=\\"#1DDF93\\" /><stop offset=\\".484\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><symbol id=\\"twitter\\" viewBox=\\"0 0 273.4 222.2\\">
<path d=\\"M273.4,26.3c-10.1,4.5-20.9,7.5-32.2,8.8c11.6-6.9,20.5-17.9,24.7-31C255,10.5,243,15.2,230.2,17.7
	C220,6.8,205.4,0,189.3,0c-31,0-56.1,25.1-56.1,56.1c0,4.4,0.5,8.7,1.5,12.8C88,66.5,46.7,44.2,19,10.3c-4.8,8.3-7.6,17.9-7.6,28.2
	c0,19.5,9.9,36.6,25,46.7c-9.2-0.3-17.8-2.8-25.4-7c0,0.2,0,0.5,0,0.7c0,27.2,19.3,49.8,45,55c-4.7,1.3-9.7,2-14.8,2
	c-3.6,0-7.1-0.4-10.6-1c7.1,22.3,27.9,38.5,52.4,39c-19.2,15-43.4,24-69.7,24c-4.5,0-9-0.3-13.4-0.8c24.8,15.9,54.3,25.2,86,25.2
	c103.2,0,159.6-85.5,159.6-159.6c0-2.4-0.1-4.9-0.2-7.3C256.4,47.4,265.9,37.5,273.4,26.3z\\" />
</symbol></defs><use width=\\"70\\" height=\\"70\\" xlink:href=\\"#appcode\\" transform=\\"translate(0, 0)\\" /><use width=\\"273.4\\" height=\\"222.2\\" xlink:href=\\"#twitter\\" transform=\\"translate(0, 80)\\" /></svg>
  </div>


<svg>
  <use xlink:href=\\"#twitter\\"></use>
</svg>

<svg>
  <use xlink:href=\\"#appcode\\"></use>
</svg>

<script type=\\"text/javascript\\" src=\\"main.js\\"></script></body>
</html>"
`;

exports[`extract-svg-sprite-webpack-plugin should works with extracted CSS: main.css 1`] = `
".a {
  background: url(sprite.svg) no-repeat 0 88.89%;
  background-size: 100% 140.50%;
}

.b {
  background: url(sprite.svg) no-repeat 0 0;
  background-size: 390.57% 446%;
}"
`;

exports[`extract-svg-sprite-webpack-plugin should works with extracted CSS: sprite.svg 1`] = `
"<svg width=\\"273.4\\" height=\\"312.2\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"><defs><symbol viewBox=\\"0 0 70 70\\" id=\\"appcode\\"><path fill=\\"#247CE6\\" d=\\"M59.2 55L70 26.3 37.3 19 33 31.2z\\" /><path fill=\\"url(#appcode_a)\\" d=\\"M70 56.1L53.5 70 20.6 61.5 29.4 45.5z\\" /><path fill=\\"url(#appcode_b)\\" d=\\"M9.7 34.8L1.3 10.8 38.4 0 35.4 31z\\" /><path fill=\\"url(#appcode_c)\\" d=\\"M61.1 40.5L50.7 22.7 50.9 22.6 38.4 0 0 41.5 0 70 69.8 56.1z\\" /><g><path fill=\\"#000\\" d=\\"M13.4 13.4H56.6V56.6H13.4z\\" /><path fill=\\"#FFF\\" d=\\"M17.4 48.5H33.599999999999994V51.2H17.4z\\" /><path d=\\"M24.5,19h3.6l7.6,17.9h-4.1l-1.6-4h-7.5l-1.6,4h-4L24.5,19z M28.6,29.4l-2.4-5.8l-2.4,5.8H28.6z\\" fill=\\"#FFF\\" /><path d=\\"M34.7,28.1L34.7,28.1c0-5.2,3.9-9.4,9.4-9.4c3.4,0,5.4,1.1,7.1,2.8l-2.5,2.9c-1.4-1.3-2.8-2-4.6-2 c-3,0-5.2,2.5-5.2,5.6V28c0,3.1,2.1,5.7,5.2,5.7c2.1,0,3.3-0.8,4.7-2.1l2.5,2.5c-1.9,2-3.9,3.2-7.4,3.2 C38.6,37.3,34.7,33.2,34.7,28.1\\" fill=\\"#FFF\\" /></g></symbol><linearGradient id=\\"appcode_a\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"20.617\\" y1=\\"57.755\\" x2=\\"70\\" y2=\\"57.755\\"><stop offset=\\".194\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><linearGradient id=\\"appcode_b\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"1.274\\" y1=\\"17.415\\" x2=\\"38.41\\" y2=\\"17.415\\"><stop offset=\\".194\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><linearGradient id=\\"appcode_c\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"45.876\\" y1=\\"72.222\\" x2=\\"11.197\\" y2=\\"23.824\\"><stop offset=\\".091\\" stop-color=\\"#1DDF93\\" /><stop offset=\\".484\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><symbol id=\\"twitter\\" viewBox=\\"0 0 273.4 222.2\\">
<path d=\\"M273.4,26.3c-10.1,4.5-20.9,7.5-32.2,8.8c11.6-6.9,20.5-17.9,24.7-31C255,10.5,243,15.2,230.2,17.7
	C220,6.8,205.4,0,189.3,0c-31,0-56.1,25.1-56.1,56.1c0,4.4,0.5,8.7,1.5,12.8C88,66.5,46.7,44.2,19,10.3c-4.8,8.3-7.6,17.9-7.6,28.2
	c0,19.5,9.9,36.6,25,46.7c-9.2-0.3-17.8-2.8-25.4-7c0,0.2,0,0.5,0,0.7c0,27.2,19.3,49.8,45,55c-4.7,1.3-9.7,2-14.8,2
	c-3.6,0-7.1-0.4-10.6-1c7.1,22.3,27.9,38.5,52.4,39c-19.2,15-43.4,24-69.7,24c-4.5,0-9-0.3-13.4-0.8c24.8,15.9,54.3,25.2,86,25.2
	c103.2,0,159.6-85.5,159.6-159.6c0-2.4-0.1-4.9-0.2-7.3C256.4,47.4,265.9,37.5,273.4,26.3z\\" />
</symbol></defs><use width=\\"70\\" height=\\"70\\" xlink:href=\\"#appcode\\" transform=\\"translate(0, 0)\\" /><use width=\\"273.4\\" height=\\"222.2\\" xlink:href=\\"#twitter\\" transform=\\"translate(0, 80)\\" /></svg>"
`;

exports[`should share sprite data with html-webpack-plugin: index.html 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"UTF-8\\">
  <title>Webpack App</title>
</head>
<body>

  <div style=\\"position: absolute; width: 0; height: 0; overflow: hidden;\\">
    <svg width=\\"273.4\\" height=\\"312.2\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"><defs><symbol viewBox=\\"0 0 70 70\\" id=\\"appcode\\"><path fill=\\"#247CE6\\" d=\\"M59.2 55L70 26.3 37.3 19 33 31.2z\\" /><path fill=\\"url(#appcode_a)\\" d=\\"M70 56.1L53.5 70 20.6 61.5 29.4 45.5z\\" /><path fill=\\"url(#appcode_b)\\" d=\\"M9.7 34.8L1.3 10.8 38.4 0 35.4 31z\\" /><path fill=\\"url(#appcode_c)\\" d=\\"M61.1 40.5L50.7 22.7 50.9 22.6 38.4 0 0 41.5 0 70 69.8 56.1z\\" /><g><path fill=\\"#000\\" d=\\"M13.4 13.4H56.6V56.6H13.4z\\" /><path fill=\\"#FFF\\" d=\\"M17.4 48.5H33.599999999999994V51.2H17.4z\\" /><path d=\\"M24.5,19h3.6l7.6,17.9h-4.1l-1.6-4h-7.5l-1.6,4h-4L24.5,19z M28.6,29.4l-2.4-5.8l-2.4,5.8H28.6z\\" fill=\\"#FFF\\" /><path d=\\"M34.7,28.1L34.7,28.1c0-5.2,3.9-9.4,9.4-9.4c3.4,0,5.4,1.1,7.1,2.8l-2.5,2.9c-1.4-1.3-2.8-2-4.6-2 c-3,0-5.2,2.5-5.2,5.6V28c0,3.1,2.1,5.7,5.2,5.7c2.1,0,3.3-0.8,4.7-2.1l2.5,2.5c-1.9,2-3.9,3.2-7.4,3.2 C38.6,37.3,34.7,33.2,34.7,28.1\\" fill=\\"#FFF\\" /></g></symbol><linearGradient id=\\"appcode_a\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"20.617\\" y1=\\"57.755\\" x2=\\"70\\" y2=\\"57.755\\"><stop offset=\\".194\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><linearGradient id=\\"appcode_b\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"1.274\\" y1=\\"17.415\\" x2=\\"38.41\\" y2=\\"17.415\\"><stop offset=\\".194\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><linearGradient id=\\"appcode_c\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"45.876\\" y1=\\"72.222\\" x2=\\"11.197\\" y2=\\"23.824\\"><stop offset=\\".091\\" stop-color=\\"#1DDF93\\" /><stop offset=\\".484\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><symbol id=\\"twitter\\" viewBox=\\"0 0 273.4 222.2\\">
<path d=\\"M273.4,26.3c-10.1,4.5-20.9,7.5-32.2,8.8c11.6-6.9,20.5-17.9,24.7-31C255,10.5,243,15.2,230.2,17.7
	C220,6.8,205.4,0,189.3,0c-31,0-56.1,25.1-56.1,56.1c0,4.4,0.5,8.7,1.5,12.8C88,66.5,46.7,44.2,19,10.3c-4.8,8.3-7.6,17.9-7.6,28.2
	c0,19.5,9.9,36.6,25,46.7c-9.2-0.3-17.8-2.8-25.4-7c0,0.2,0,0.5,0,0.7c0,27.2,19.3,49.8,45,55c-4.7,1.3-9.7,2-14.8,2
	c-3.6,0-7.1-0.4-10.6-1c7.1,22.3,27.9,38.5,52.4,39c-19.2,15-43.4,24-69.7,24c-4.5,0-9-0.3-13.4-0.8c24.8,15.9,54.3,25.2,86,25.2
	c103.2,0,159.6-85.5,159.6-159.6c0-2.4-0.1-4.9-0.2-7.3C256.4,47.4,265.9,37.5,273.4,26.3z\\" />
</symbol></defs><use width=\\"70\\" height=\\"70\\" xlink:href=\\"#appcode\\" transform=\\"translate(0, 0)\\" /><use width=\\"273.4\\" height=\\"222.2\\" xlink:href=\\"#twitter\\" transform=\\"translate(0, 80)\\" /></svg>
  </div>


<svg>
  <use xlink:href=\\"#twitter\\"></use>
</svg>

<svg>
  <use xlink:href=\\"#appcode\\"></use>
</svg>

<script type=\\"text/javascript\\" src=\\"main.js\\"></script></body>
</html>"
`;

exports[`should works with extracted CSS: main.css 1`] = `
".a {
  background: url(sprite.svg) no-repeat 0 88.89%;
  background-size: 100% 140.50%;
}

.b {
  background: url(sprite.svg) no-repeat 0 0;
  background-size: 390.57% 446%;
}"
`;

exports[`should works with extracted CSS: sprite.svg 1`] = `
"<svg width=\\"273.4\\" height=\\"312.2\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\"><defs><symbol viewBox=\\"0 0 70 70\\" id=\\"appcode\\"><path fill=\\"#247CE6\\" d=\\"M59.2 55L70 26.3 37.3 19 33 31.2z\\" /><path fill=\\"url(#appcode_a)\\" d=\\"M70 56.1L53.5 70 20.6 61.5 29.4 45.5z\\" /><path fill=\\"url(#appcode_b)\\" d=\\"M9.7 34.8L1.3 10.8 38.4 0 35.4 31z\\" /><path fill=\\"url(#appcode_c)\\" d=\\"M61.1 40.5L50.7 22.7 50.9 22.6 38.4 0 0 41.5 0 70 69.8 56.1z\\" /><g><path fill=\\"#000\\" d=\\"M13.4 13.4H56.6V56.6H13.4z\\" /><path fill=\\"#FFF\\" d=\\"M17.4 48.5H33.599999999999994V51.2H17.4z\\" /><path d=\\"M24.5,19h3.6l7.6,17.9h-4.1l-1.6-4h-7.5l-1.6,4h-4L24.5,19z M28.6,29.4l-2.4-5.8l-2.4,5.8H28.6z\\" fill=\\"#FFF\\" /><path d=\\"M34.7,28.1L34.7,28.1c0-5.2,3.9-9.4,9.4-9.4c3.4,0,5.4,1.1,7.1,2.8l-2.5,2.9c-1.4-1.3-2.8-2-4.6-2 c-3,0-5.2,2.5-5.2,5.6V28c0,3.1,2.1,5.7,5.2,5.7c2.1,0,3.3-0.8,4.7-2.1l2.5,2.5c-1.9,2-3.9,3.2-7.4,3.2 C38.6,37.3,34.7,33.2,34.7,28.1\\" fill=\\"#FFF\\" /></g></symbol><linearGradient id=\\"appcode_a\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"20.617\\" y1=\\"57.755\\" x2=\\"70\\" y2=\\"57.755\\"><stop offset=\\".194\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><linearGradient id=\\"appcode_b\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"1.274\\" y1=\\"17.415\\" x2=\\"38.41\\" y2=\\"17.415\\"><stop offset=\\".194\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><linearGradient id=\\"appcode_c\\" gradientUnits=\\"userSpaceOnUse\\" x1=\\"45.876\\" y1=\\"72.222\\" x2=\\"11.197\\" y2=\\"23.824\\"><stop offset=\\".091\\" stop-color=\\"#1DDF93\\" /><stop offset=\\".484\\" stop-color=\\"#00DAF0\\" /><stop offset=\\".903\\" stop-color=\\"#247CE6\\" /></linearGradient><symbol id=\\"twitter\\" viewBox=\\"0 0 273.4 222.2\\">
<path d=\\"M273.4,26.3c-10.1,4.5-20.9,7.5-32.2,8.8c11.6-6.9,20.5-17.9,24.7-31C255,10.5,243,15.2,230.2,17.7
	C220,6.8,205.4,0,189.3,0c-31,0-56.1,25.1-56.1,56.1c0,4.4,0.5,8.7,1.5,12.8C88,66.5,46.7,44.2,19,10.3c-4.8,8.3-7.6,17.9-7.6,28.2
	c0,19.5,9.9,36.6,25,46.7c-9.2-0.3-17.8-2.8-25.4-7c0,0.2,0,0.5,0,0.7c0,27.2,19.3,49.8,45,55c-4.7,1.3-9.7,2-14.8,2
	c-3.6,0-7.1-0.4-10.6-1c7.1,22.3,27.9,38.5,52.4,39c-19.2,15-43.4,24-69.7,24c-4.5,0-9-0.3-13.4-0.8c24.8,15.9,54.3,25.2,86,25.2
	c103.2,0,159.6-85.5,159.6-159.6c0-2.4-0.1-4.9-0.2-7.3C256.4,47.4,265.9,37.5,273.4,26.3z\\" />
</symbol></defs><use width=\\"70\\" height=\\"70\\" xlink:href=\\"#appcode\\" transform=\\"translate(0, 0)\\" /><use width=\\"273.4\\" height=\\"222.2\\" xlink:href=\\"#twitter\\" transform=\\"translate(0, 80)\\" /></svg>"
`;
